scripts/docs_link_checker.py

```diff
@@ -0,0 +1,306 @@
+        else:
+            logger.debug("Relative image %s in file %s found", link, file)
+            return None
+
+    def _check_docroot_link(self, link: str, file: str, path: str) -> LinkReport:
+        logger.debug("Checking docroot link %s in file %s", link, file)
+
+        md_file = self._get_docroot_path(path)
+        if not os.path.isfile(md_file):
+            logger.info("Docroot link %s in file %s was not found", link, file)
+            return LinkReport(link, file, f"Linked file {md_file} not found")
+        else:
+            logger.debug("Docroot link %s in file %s found", link, file)
+            return None
+
+    def check_link(self, match: re.Match, file: str) -> LinkReport:
+        """Checks that a link is valid. Valid links are:
+        - Absolute links that begin with a forward slash and the specified site prefix (ex: /docs) with no suffix
+        - Absolute images with an image suffix
+        - Relative links that begin with either . or .. and have a .md suffix
+        - Relative images with an image suffix
+        - Docroot links that begin with a character (neither . or /) are relative to the doc root (ex: /docs) and have a .md suffix
+
+        Args:
+            match: A positive match of a markdown link (ex: [...](...))
+            file: The file where the match was found
+
+        Returns:
+            A LinkReport if the link is broken, otherwise None
+        """
+        link = match.group(2)
+
+        # skip links that are anchor only (start with #)
+        if self._is_anchor_link(link):
+            return None
+
+        if self._external_link_pattern.match(link):
+            result = self._check_external_link(link, file)
+        elif self._is_image_link(match.group(0)):
+            match = self._relative_image_pattern.match(link)
+            if match:
+                result = self._check_relative_image(link, file, match.group("path"))
+            else:
+                match = self._absolute_image_pattern.match(link)
+                if match:
+                    result = self._check_absolute_image(link, file, match.group("path"))
+                else:
+                    result = LinkReport(link, file, "Invalid image link format")
+        else:
+            match = self._relative_link_pattern.match(link)
+            if match:
+                result = self._check_relative_link(link, file, match.group("path"))
+            else:
+                match = self._absolute_link_pattern.match(link)
+                if match:
+                    result = self._check_absolute_link(link, file, match.group("path"))
+                else:
+                    match = self._docroot_link_pattern.match(link)
+                    if match:
+                        result = self._check_docroot_link(
+                            link, file, match.group("path")
+                        )
+                    else:
+                        result = LinkReport(link, file, "Invalid link format")
+
```

Comment:
Couldn't think of a super clear way to refactor this but since this isn't GE code, I'm happy to leave it as is.

Reply:
I could have used a strategy pattern here (and when I got to my 4th nested if, I probably should have) ... but it's a very linear script so I decided to forego the effort.